types/src/v0/impls/l1.rs

```diff
@@ -88,7 +110,73 @@ impl JsonRpcClient for RpcClient {
-            Self::Ws(client) => client.request(method, params).await?,
+            Self::Ws {
+                conn,
+                reconnect,
+                url,
+                retry_delay,
+            } => {
+                let conn_guard = conn
+                    .try_read()
+                    // We only lock the connection exclusively when we are resetting it, so if it is
+                    // locked that means it was closed and is still being reset. There is no point
+                    // in trying a request with a closed connection.
+                    .map_err(|_| {
+                        ProviderError::CustomError("connection closed; reset in progress".into())
+                    })?;
+                match conn_guard.request(method, params).await {
+                    Ok(res) => res,
+                    Err(err @ WsClientError::UnexpectedClose) => {
+                        // If the WebSocket connection is closed, try to reopen it.
+                        if let Ok(mut reconnect_guard) = reconnect.try_lock() {
+                            if matches!(*reconnect_guard, L1ReconnectTask::Idle) {
+                                // No one is currently resetting this connection, so it's up to us.
+                                let conn = conn.clone();
+                                let reconnect = reconnect.clone();
+                                let url = url.clone();
+                                let retry_delay = *retry_delay;
+                                let span = tracing::warn_span!("ws resetter");
+                                *reconnect_guard = L1ReconnectTask::Reconnecting(spawn(
+                                    async move {
+                                        tracing::warn!("ws connection closed, trying to reset");
+                                        let new_conn = loop {
+                                            match Ws::connect(url.clone()).await {
+                                                Ok(conn) => break conn,
+                                                Err(err) => {
+                                                    tracing::warn!("failed to reconnect: {err:#}");
+                                                    sleep(retry_delay).await;
+                                                }
+                                            }
+                                        };
+
+                                        // Reset the connection, and set the reconnect task back to
+                                        // idle, so that the connection can be reset again if
+                                        // needed.
+                                        let mut conn = conn.write().await;
+                                        let mut reconnect = reconnect.lock().await;
+                                        *conn = new_conn;
+                                        if !matches!(*reconnect, L1ReconnectTask::Cancelled) {
```

Comment:
What does L1ReconnectTask::Cancelled exactly mean? We are spinning up the reconnect task here too and updating the connection